src/edu/washington/escience/myria/operator/UniqueTupleHashTable.java

```diff
@@ -0,0 +1,167 @@
+  private static final long serialVersionUID = 1L;
+  /**
+   * We store this value instead of a valid index to indicate
+   * that a given hash code is mapped to multiple indexes.
+   */
+  private static final int COLLIDING_KEY = -1;
+  /**
+   * We return this value from getIfAbsent() to indicate absence,
+   * since 0 and -1 are already legitimate values.
+   */
+  private static final int ABSENT_VALUE = -2;
+
+  /** Map from unique hash codes to indexes. */
+  private transient LongIntHashMap keyHashCodesToIndexes;
+  /** Map from colliding hash codes to indexes. */
+  private transient LongObjectHashMap<IntArrayList> collidingKeyHashCodesToIndexes;
+  /** The table containing keys and values. */
+  private transient MutableTupleBuffer data;
+  /** Key column indices. */
+  private final int[] keyColumns;
+
+  /** The logger for this class. */
+  protected static final org.slf4j.Logger LOGGER =
+      LoggerFactory.getLogger(UniqueTupleHashTable.class);
+
+  /**
+   * @param schema schema
+   * @param keyColumns key column indices
+   */
+  public UniqueTupleHashTable(final Schema schema, final int[] keyColumns) {
+    this.keyColumns = keyColumns;
+    data = new MutableTupleBuffer(schema);
+    keyHashCodesToIndexes = new LongIntHashMap();
+    collidingKeyHashCodesToIndexes = new LongObjectHashMap<IntArrayList>();
+  }
+
+  /**
+   * @return the number of tuples this hash table has.
+   */
+  public int numTuples() {
+    return data.numTuples();
+  }
+
+  /**
+   * Get the data table index given key columns from a tuple in a tuple batch.
+   *
+   * @param tb the input tuple batch
+   * @param key the key columns
+   * @param row the row index of the tuple
+   * @return the index of the matching tuple in the data table, or -1 if no match
+   */
+  public int getIndex(final ReadableTable tb, final int[] key, final int row) {
+    final long hashcode = HashUtils.hashSubRowLong(tb, key, row);
+    int index = keyHashCodesToIndexes.getIfAbsent(hashcode, ABSENT_VALUE);
+    if (index == ABSENT_VALUE) {
+      return -1;
+    }
+    if (index == COLLIDING_KEY) {
+      IntArrayList collidingIndexes = collidingKeyHashCodesToIndexes.get(hashcode);
+      Preconditions.checkNotNull(collidingIndexes);
+      Preconditions.checkState(collidingIndexes.size() > 1);
+      IntIterator iter = collidingIndexes.intIterator();
+      while (iter.hasNext()) {
+        int idx = iter.next();
+        if (TupleUtils.tupleEquals(tb, key, row, data, keyColumns, idx)) {
+          return idx;
+        }
+      }
+      assert false; // we should never get here
+    }
+    return index;
+  }
+
+  /**
+   * Replace a matching tuple in the data table with the input tuple.
+   *
+   * @param tb the input tuple batch
+   * @param keyColumns the key columns
+   * @param row the row index of the input tuple
+   * @return if at least one tuple is replaced
+   */
+  public boolean replace(final TupleBatch tb, final int[] keyColumns, final int row) {
```

Comment:
Right, I just duplicated it from `TupleHashTable` because there didn't seem to be any reason not to include it in the interface. (Arguably, they should both be explicitly implementing interfaces deriving from a common interface, which differ only on `getIndices()` vs. `getIndex()`.)

I prefer to keep the semantics of `UniqueTupleHashTable` clear in the interface (i.e., have only a `getIndex()` method), although certainly the overflow table idea would also work well when multiple tuples for a grouping key are rare but possible. I think we'd need to do a bit more refactoring in that case (probably factor out the `UniqueTupleHashTable` implementation and have a wrapper that implements the `getIndices()` interface and delegates to this implementation).